### PR TITLE
Add allowed scopes to introspect response after scope validation

### PR DESCRIPTION
**Pull request by sumedhe**
### Proposed changes in this pull request

Add allowed scopes to introspect response and to the accessTokenDO after scope validation.

### Related issue
wso2/product-is#14429

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java [hunk 1]
```diff
@@ -559,6 +559,10 @@ private OAuth2IntrospectionResponseDTO validateAccessToken(OAuth2TokenValidation
         }
 
         addAllowedScopes(messageContext, requestedAllowedScopes.toArray(new String[0]));
+        if (accessTokenDO != null) {
+            addAllowedScopes(introResp, accessTokenDO, requestedAllowedScopes.toArray(new String[0]));
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java:563]: shall we add a comment what is the difference between these two addAllowedScopes methods?

Comment (sumedhe) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java:566]: Updated

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java [hunk 2]
```diff
@@ -753,4 +760,13 @@ private void addAllowedScopes(OAuth2TokenValidationMessageContext oAuth2TokenVal
         String[] scopesToReturn = (String[]) ArrayUtils.addAll(scopes, allowedScopes);
         oAuth2TokenValidationMessageContext.getResponseDTO().setScope(scopesToReturn);
     }
+
+    private void addAllowedScopes(OAuth2IntrospectionResponseDTO introResp, AccessTokenDO accessTokenDO,
+                                  String[] allowedScopes) {
+
+        String[] scopes = accessTokenDO.getScope();
+        String[] scopesToReturn = (String[]) ArrayUtils.addAll(scopes, allowedScopes);
+        accessTokenDO.setScope(scopesToReturn);
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java:769]: no need to set to accessTokenDO as we already done the scope validation with the required scopes.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java [hunk 3]
```diff
@@ -753,4 +760,13 @@ private void addAllowedScopes(OAuth2TokenValidationMessageContext oAuth2TokenVal
         String[] scopesToReturn = (String[]) ArrayUtils.addAll(scopes, allowedScopes);
         oAuth2TokenValidationMessageContext.getResponseDTO().setScope(scopesToReturn);
     }
+
+    private void addAllowedScopes(OAuth2IntrospectionResponseDTO introResp, AccessTokenDO accessTokenDO,
+                                  String[] allowedScopes) {
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java:765]: change the allowedScopes -> requestedAllowedScopes

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java [hunk 4]
```diff
@@ -753,4 +760,13 @@ private void addAllowedScopes(OAuth2TokenValidationMessageContext oAuth2TokenVal
...
+
+        String[] scopes = accessTokenDO.getScope();
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java:767]: change the name scopes->validatedScopeArray

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java [hunk 5]
```diff
@@ -753,4 +760,13 @@ private void addAllowedScopes(OAuth2TokenValidationMessageContext oAuth2TokenVal
         String[] scopesToReturn = (String[]) ArrayUtils.addAll(scopes, allowedScopes);
         oAuth2TokenValidationMessageContext.getResponseDTO().setScope(scopesToReturn);
     }
+
+    private void addAllowedScopes(OAuth2IntrospectionResponseDTO introResp, AccessTokenDO accessTokenDO,
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java:764]: Change the method name to addScopeToIntrospectionResponse

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java (hunk 2 above)
Comment (sumedhe) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java]: fixed with f5642b437c1e3a57b5c669dbc18b32d53d613f68

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java (hunk 3 above)
Comment (sumedhe) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java]: fixed with f5642b437c1e3a57b5c669dbc18b32d53d613f68

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java (hunk 5 above)
Comment (sumedhe) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java]: fixed with f5642b437c1e3a57b5c669dbc18b32d53d613f68

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java (hunk 4 above)
Comment (sumedhe) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java]: changed to `validatedScopes`

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java [hunk 6]
```diff
@@ -558,7 +558,14 @@ private OAuth2IntrospectionResponseDTO validateAccessToken(OAuth2TokenValidation
             return buildIntrospectionErrorResponse("Scope validation failed");
         }
 
+        // Add requested allowed scopes to the message context.
         addAllowedScopes(messageContext, requestedAllowedScopes.toArray(new String[0]));
+
+        // Add requested allowed scopes to introResp and to the accessTokenDO.
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/TokenValidationHandler.java:564]:
```suggestion
        // Add requested allowed scopes and validated scopes to introResp.
```

**Review by nilasini — APPROVED** [review]

**Pull request merged**